src/main/java/com/aws/greengrass/logmanager/model/LogFileGroup.java

```diff
@@ -1,75 +1,120 @@
+     * @return list of logFile.
+     * @throws InvalidLogGroupException the exception if this is not a valid directory.
+     */
+    public static LogFileGroup create(Pattern filePattern, URI directoryURI, Instant lastUpdated,
+                                      Path hardLinksDirectory)
             throws InvalidLogGroupException {
         File folder = new File(directoryURI);
-        fileHashToLogFile = new ConcurrentHashMap<>();
 
         if (!folder.isDirectory()) {
             throw new InvalidLogGroupException(String.format("%s must be a directory", directoryURI));
         }
 
-        LogFile[] files = LogFile.of(folder.listFiles());
+        Path componentHardlinksDirectory = hardLinksDirectory.resolve(filePattern.toString());
+
+        try {
+            Utils.deleteFileRecursively(componentHardlinksDirectory.toFile());
+            Utils.createPaths(componentHardlinksDirectory);
+        } catch (IOException e) {
+            throw new InvalidLogGroupException(
+                    String.format("%s failed to create hard link directory", componentHardlinksDirectory), e);
+        }
+
+        File[] files = folder.listFiles();
+        if (files == null) {
+            logger.atWarn().log("logFiles is null");
+            return new LogFileGroup(Collections.emptyList(), filePattern, new HashMap<>());
+        }
+
+        Map<String, LogFile> fileHashToLogFile = new ConcurrentHashMap<>();
         List<LogFile> allFiles = new ArrayList<>();
-        if (files.length != 0) {
-            for (LogFile file: files) {
-                String fileHash = file.hashString();
-                boolean isModifiedAfterLastUpdatedFile =
-                        lastUpdated.isBefore(Instant.ofEpochMilli(file.lastModified()));
-                boolean isNameMatchPattern = filePattern.matcher(file.getName()).find();
-                boolean isEmptyFileHash = Utils.isEmpty(fileHash);
-
-                if (file.isFile()
-                        && isModifiedAfterLastUpdatedFile
-                        && isNameMatchPattern
-                        && !isEmptyFileHash) {
-                    allFiles.add(file);
-                    fileHashToLogFile.put(fileHash, file);
+        // TODO: We have to add the rotation detection mechanism here otherwise there is a chance that while we are
+        //  looping and creating the hardlinks the files gets rotated so the path that
+        for (File file : files) {
+            boolean isModifiedAfterLastUpdatedFile =
+                    lastUpdated.isBefore(Instant.ofEpochMilli(file.lastModified()));
+            boolean isNameMatchPattern = filePattern.matcher(file.getName()).find();
+
+            if (file.isFile() && isModifiedAfterLastUpdatedFile && isNameMatchPattern) {
+                try {
+                    LogFile logfile = LogFile.of(file, componentHardlinksDirectory);
+                    allFiles.add(logfile);
+                    fileHashToLogFile.put(logfile.hashString(), logfile);
```

Comment:
+1 for calculating hash after the other checks.

Reply:
Don't know what this comment means but the hardlink gets memoized after the first time we calculate it to avoid additional reads